L1Trigger/L1TTrackMatch/src/L1TkMuMantra.cc

```diff
@@ -0,0 +1,214 @@
+        LogTrace("L1TkMuMantra") << "... fit phi cent : " << fc << std::endl;
+        LogTrace("L1TkMuMantra") << "... fit phi high : " << fh << std::endl;
+      }
+      throw std::runtime_error("L1TkMuMantra : Could not init phi");
+    }
+    wdws_phi_.at(ib).SetName(wdn);
+    wdws_phi_.at(ib).SetLower(fl);
+    wdws_phi_.at(ib).SetCentral(fc);
+    wdws_phi_.at(ib).SetUpper(fh);
+  }
+}
+
+int L1TkMuMantra::findBin(double val) {
+  // FIXME: not the most efficient, nor the most elegant implementation for now
+  if (val < bounds_.at(0))
+    return 0;
+  if (val >= bounds_.back())
+    return (nbins_ - 1);  // i.e. bounds_size() -2
+
+  for (uint ib = 0; ib < bounds_.size() - 1; ++ib) {
+    if (val >= bounds_.at(ib) && val < bounds_.at(ib + 1))
+      return ib;
+  }
+
+  if (verbosity_ > 0)
+    LogTrace("L1TkMuMantra") << "Something strange happened at val " << val << std::endl;
+  return 0;
+}
+
+void L1TkMuMantra::test(double eta, double pt) {
+  int ibin = findBin(eta);
+
+  LogTrace("L1TkMuMantra") << " ---- eta : " << eta << " pt: " << pt << std::endl;
+  LogTrace("L1TkMuMantra") << " ---- bin " << ibin << std::endl;
+  LogTrace("L1TkMuMantra") << " ---- "
+                           << "- low_phi  : " << wdws_phi_.at(ibin).bound_low(pt)
+                           << "- cent_phi : " << wdws_phi_.at(ibin).bound_cent(pt)
+                           << "- high_phi : " << wdws_phi_.at(ibin).bound_high(pt) << std::endl;
+
+  LogTrace("L1TkMuMantra") << " ---- "
+                           << "- low_theta  : " << wdws_theta_.at(ibin).bound_low(pt)
+                           << "- cent_theta : " << wdws_theta_.at(ibin).bound_cent(pt)
+                           << "- high_theta : " << wdws_theta_.at(ibin).bound_high(pt) << std::endl;
+
+  return;
+}
+
+std::vector<int> L1TkMuMantra::find_match(std::vector<track_df>& tracks, std::vector<muon_df>& muons) {
+  std::vector<int> result(muons.size(), -1);  // init all TkMu to index -1
+  for (uint imu = 0; imu < muons.size(); ++imu) {
+    muon_df mu = muons.at(imu);
+    std::vector<std::pair<double, int>> matched_trks;  // sort_par, idx
+    for (uint itrk = 0; itrk < tracks.size(); ++itrk) {
+      // preselection of tracks
+      track_df trk = tracks.at(itrk);
+      if (trk.chi2 >= max_chi2)
+        continue;  // require trk.chi2 < max_chi2
+      if (trk.nstubs < min_nstubs)
+        continue;  // require trk.nstubs >= min_nstubs
+
+      // compute delta
+      // col_dphicharge = dataframe.loc[ : , 'trk_phi'].subtract(dataframe.loc[ : , phi_name]).multiply(dataframe.loc[ : , 'trk_charge']).apply(to_mpi_pi)
+      // col_dthetaendc = dataframe.loc[ : , theta_name].subtract(dataframe.loc[ : , 'trk_theta']).multiply(dataframe.loc[ : , etasign_name])
+
+      double dphi_charge = to_mpi_pi((trk.phi - mu.phi) * trk.charge);
+      double dtheta_endc =
+          (mu.theta - trk.theta) * (std::abs(mu.theta) / mu.theta);  // sign from theta, to avoid division by 0
```

Comment:
you have a sign function defined, use it

Reply:
done